pilot/cmd/pilot-discovery/main.go

```diff
@@ -105,89 +106,91 @@ func init() {
 		"Enable profiling via web interface host:port/debug/pprof")
 
 	// Use TLS certificates if provided.
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.ServerOptions.TLSOptions.CaCertFile, "caCertFile", "",
+	c.PersistentFlags().StringVar(&serverArgs.ServerOptions.TLSOptions.CaCertFile, "caCertFile", "",
 		"File containing the x509 Server CA Certificate")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.ServerOptions.TLSOptions.CertFile, "tlsCertFile", "",
+	c.PersistentFlags().StringVar(&serverArgs.ServerOptions.TLSOptions.CertFile, "tlsCertFile", "",
 		"File containing the x509 Server Certificate")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.ServerOptions.TLSOptions.KeyFile, "tlsKeyFile", "",
+	c.PersistentFlags().StringVar(&serverArgs.ServerOptions.TLSOptions.KeyFile, "tlsKeyFile", "",
 		"File containing the x509 private key matching --tlsCertFile")
-	discoveryCmd.PersistentFlags().StringSliceVar(&serverArgs.ServerOptions.TLSOptions.TLSCipherSuites, "tls-cipher-suites", nil,
+	c.PersistentFlags().StringSliceVar(&serverArgs.ServerOptions.TLSOptions.TLSCipherSuites, "tls-cipher-suites", nil,
 		"Comma-separated list of cipher suites for istiod TLS server. "+
 			"If omitted, the default Go cipher suites will be used. \n"+
 			"Preferred values: "+strings.Join(secureTLSCipherNames(), ", ")+". \n"+
 			"Insecure values: "+strings.Join(insecureTLSCipherNames(), ", ")+".")
 
-	discoveryCmd.PersistentFlags().Float32Var(&serverArgs.RegistryOptions.KubeOptions.KubernetesAPIQPS, "kubernetesApiQPS", 80.0,
+	c.PersistentFlags().Float32Var(&serverArgs.RegistryOptions.KubeOptions.KubernetesAPIQPS, "kubernetesApiQPS", 80.0,
 		"Maximum QPS when communicating with the kubernetes API")
 
-	discoveryCmd.PersistentFlags().IntVar(&serverArgs.RegistryOptions.KubeOptions.KubernetesAPIBurst, "kubernetesApiBurst", 160,
+	c.PersistentFlags().IntVar(&serverArgs.RegistryOptions.KubeOptions.KubernetesAPIBurst, "kubernetesApiBurst", 160,
 		"Maximum burst for throttle when communicating with the kubernetes API")
 
 	// Attach the Istio logging options to the command.
-	loggingOptions.AttachCobraFlags(rootCmd)
+	loggingOptions.AttachCobraFlags(c)
 
 	// Attach the Istio Ctrlz options to the command.
-	serverArgs.CtrlZOptions.AttachCobraFlags(rootCmd)
+	serverArgs.CtrlZOptions.AttachCobraFlags(c)
 
 	// Attach the Istio Keepalive options to the command.
-	serverArgs.KeepaliveOptions.AttachCobraFlags(rootCmd)
-
-	cmd.AddFlags(rootCmd)
+	serverArgs.KeepaliveOptions.AttachCobraFlags(c)
+}
 
+func main() {
+	rootCmd := newRootCommand()
+	discoveryCmd := newDiscoveryCommand()
+	addFlags(discoveryCmd)
 	rootCmd.AddCommand(discoveryCmd)
 	rootCmd.AddCommand(version.CobraCommand())
 	rootCmd.AddCommand(collateral.CobraCommand(rootCmd, &doc.GenManHeader{
 		Title:   "Istio Pilot Discovery",
 		Section: "pilot-discovery CLI",
 		Manual:  "Istio Pilot Discovery",
 	}))
-}
+	rootCmd.AddCommand(requestCmd)
```

Comment:
this doesn't seem consistent with other usages of Cobra in istio or others. Typically i have seen the rootCmd fully constructed <somewhere> and main just calls rootCmd.Execute().

This allows other packages to get the full constructed command and use outside of main() so it can be tested or reused

Reply:
This is not the truth, ut can not test the whole command running. 

The init function is not encouraged in golang, its running order depend on package load.

The best practice of k8s FYI (https://github.com/kubernetes/kubernetes/blob/46b9ccd0fcc9522342807b16d750b02e34998d64/cmd/kubeadm/app/cmd/cmd.go#L83-L92) has no such usage at all. In Istio we make use of `init` everywhere, especially prom register/istioctl.

The disadvantage is obvious, for ut these init functions are still running.

Reply:
I am 100% fine with removing init(). Just not running this in main(). main should just call getcmd().Execute() - just like the k8s example

Reply:
Make sense, will improve